packages/angular/ssr/test/routes/ng-routes_spec.ts

```diff
@@ -0,0 +1,166 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.dev/license
+ */
+
+/* eslint-disable import/no-unassigned-import */
+import '@angular/compiler';
+/* eslint-enable import/no-unassigned-import */
+
+import { Component } from '@angular/core';
+import { extractRoutesAndCreateRouteTree } from '../../src/routes/ng-routes';
+import { RenderMode, SSGFallback } from '../../src/routes/route-config';
+import { setAngularAppTestingManifest } from '../testing-utils';
+
+describe('extractRoutesAndCreateRouteTree', () => {
+  const url = new URL('http://localhost');
+
+  @Component({
+    standalone: true,
+    selector: 'app-dummy-comp',
+    template: `dummy works`,
+  })
+  class DummyComponent {}
+
+  it('should extract routes and create a route tree', async () => {
+    setAngularAppTestingManifest(
+      [
+        { path: 'home', component: DummyComponent },
+        { path: 'redirect', redirectTo: 'home' },
+        { path: 'user/:id', component: DummyComponent },
+      ],
+      [
+        { path: '/home', renderMode: RenderMode.CSR },
+        { path: '/redirect', renderMode: RenderMode.SSR, status: 301 },
+        { path: '/**', renderMode: RenderMode.SSR },
+      ],
+    );
+
+    const routeTree = await extractRoutesAndCreateRouteTree(url);
+    expect(routeTree.toObject()).toEqual([
+      { route: '/home', renderMode: RenderMode.CSR, redirectTo: undefined },
+      { route: '/redirect', renderMode: RenderMode.SSR, status: 301, redirectTo: '/home' },
+      { route: '/user/:id', renderMode: RenderMode.SSR, redirectTo: undefined },
+    ]);
+  });
+
+  describe('when `invokeGetPrerenderPaths` is true', () => {
+    it('should resolve parameterized routes for SSG and add a fallback route if fallback is SSR', async () => {
+      setAngularAppTestingManifest(
+        [{ path: 'user/:id/role/:role', component: DummyComponent }],
+        [
+          {
+            path: 'user/:id/role/:role',
+            renderMode: RenderMode.SSG,
+            async getPrerenderPaths() {
+              return [
+                { id: 'joe', role: 'admin' },
+                { id: 'jane', role: 'writer' },
+              ];
+            },
+          },
+          { path: '/user/:id/role/:role', renderMode: RenderMode.SSG, fallback: SSGFallback.SSR },
```

Comment:
Question: Does `fallback` need to be in its own route? Can we not move `fallback: SSGFallback.SSR` to the first route?

Reply:
We can move it into the route, but this would require special fallback rendering during the `render` phase. By adding an extra route, we can bypass this requirement and handle the fallback as a standard route instead.

Reply:
I don't think I'm following. How does having this as a separate route make the implementation easier?

From the developer's perspective, it seems weird to have a second route with a `fallback` option (isn't the second route itself the fallback?). How is this different from just using `renderMode: RenderMode.SSR` on the second route? Should it be an error that the user defined `RenderMode.SSG` and a route with two params (`:id` / `:role`), but is missing a `getPrerenderParams` implementation?

Reply:
Ohh I now understand the questions. I made a mistake in the UT.